### ISSUE TITLE
UseDotNet@2 does not use version set for build

### ISSUE THREAD

**Issue by tisis2** [microsoft/azure-pipelines-tasks#17196]
**Question, Bug, or Feature?**  
*Type*: Bug

**Enter Task Name**: [UseDotNetV2](https://github.com/microsoft/azure-pipelines-tasks/tree/master/Tasks/UseDotNetV2)

## Environment
* Azure DevOps Server 2020 Update 1.2
* Windows Agent - 2.181.2 

## Issue Description
i created a branch for our project to migrate from net6 to net7. after the first build with net7 i recognized that also our master builds run with net7 sdk now. this should not happen since there is a UseDotNet@2 Task at the beginning of the pipeline with version 6.x and the documentation clearly says "... Use this task to change the version of .NET Core used in subsequent tasks. ..."

it seems that this task only really installs the dotnet sdk version but does not take care that subsequent tasks are using the version set.

**Comment by wrakocy** [comment]
Same issue [here](https://github.com/microsoft/azure-pipelines-tasks/issues/17152).

**Comment by nilekirk** [comment]
We're seing this too, from today. And we're dependent on stuff that does not work in the presence of net7. So this is really bad - we have no working CI-builds, we can't ship stuff. And it just happened; we didn't change a thing.

**Comment by tisis2** [comment]
@nilekirk i placed a global.json in my repository root that sets sdk to 6.0.0 and only allows minor roll forward. maybe that helps as a workaround to be able to ship

**Comment by nilekirk** [comment]
@tisis2 Thanks, a global.json helped.

This is still a bug though - the task does not do what it says it does.

**Comment by tisis2** [comment]
@nilekirk yes its also a bug in my opinion... especially because one pipeline running on an agent can cause other pipelines to fail when they run later... happened that way to me... so i had to add the global.json to all my repos just because i created a branch for upgrading to net7 in one of them...

**Comment by tisis2** [comment]
nobody of microsoft even interested? any updates?

**Comment by Berthelmaster** [comment]
Any updates? This is messing with two of our Linux Agents after Installtion .NET 8 rc-2 for a project.

**Comment by willisd2** [comment]
Ran into this _exact_ issue using on-prem build agents. Created a branch to upgrade my project from .NET 6 to 7, and after a build kicked off against the branch, it started breaking pipelines for all the other branches still on .NET 6... Our build is structured where it does a UseDotNet@2 first, telling it to go off of the global.json. The next step does a `dotnet restore` on the solution which ends up using the .NET 6 SDK defined in the global.json. The next step does a `dotnet build` on the csproj's with a `--no-restore` option, which ends up ignoring the global.json and uses the .NET 7 SDK instead. This causes it to error out and fail the build.

**Comment by foomaniac** [comment]
Disappointed to have stumbled across this issue to find it's unresolved. Seems this is happening now.

Because of another bug in .net 8 preventing build steps from authenticating against an Azure DevOps private nuget feed anymore, we've been trying to find ways to allow us to upgrade to .net 8 without having to update every single one of our 100+ existing pipelines to add the auth step in.

In short, I've discovered that if we want to ensure that an agent defaults to .net 6, the only way to do this is to place a global.json at the root of the agent's folder.

![image](https://github.com/microsoft/azure-pipelines-tasks/assets/25570986/a7b9db63-a005-48a0-90d2-a42bb7f5542f)

The content of this global.json is below:

```
{
  "sdk": {
    "version": "6.0.201",
    "rollForward": "minor"
  }
}
```

We then looked to use the 'Use Dot Net SDK' task step to specify .net 8 for the relevant upgrade pipelines. Nothing in the documentation states that it wouldn't override the global.json file. https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/use-dotnet-v2?view=azure-pipelines

Infact, there's an option to tell it to use the global.json if you want to. 

Instead what is happening is, despite it successfully (apparently) setting the version to .net 8 as we're asking. Subsequent steps are still defaulting to the .net 6 version.

![image](https://github.com/microsoft/azure-pipelines-tasks/assets/25570986/59094892-2974-4e68-aec8-a8a415703428)

All appears fine, until the next dotnet cli command runs...

(BTW Yes I know in the screenshot it says .net 7, as much as microsoft can't be bothered to fix these issues, I can't be bothered going back to find the correct failed build that shows .net 8 but trust me, it still fails)
![image](https://github.com/microsoft/azure-pipelines-tasks/assets/25570986/aa2d41be-46da-4b95-b488-839a6878be79)

You can see the sdk version the dotnet command is running against is still 6, and the build fails because it isn't finding the version it expects. So the 'Use Dot Net SDK' step is not doing what it's supposed to do at all. 

We're so close to sacking off .net 8, add in the other issue we have https://github.com/dotnet/sdk/issues/37379#issuecomment-1981697406 and also this one https://github.com/dotnet/efcore/issues/25555 which is preventing us upgrading projects that use ef bundling regardless of this ticket's issue, it just makes Microsoft come across pretty poor at the moment.

**Comment by mhborg-timextender** [comment]
We have the same problem on ours, UseDotNet@2 is just being ignored in favor of the default one on our self-hosted agent.

**Comment by FRANKR318** [comment]
I have the same problem! When will this be fixed?

**Comment by L-Markham** [comment]
> I have the same problem! When will this be fixed?

It never will be. This thread has been open for a year and a half, and Microsoft has yet to reply; extremely poor behavior from them.
